### ISSUE TITLE
Performance degrades after replacing `torch.div & torch.softmax` with `Flashinfer Fused temperature online softmax kernel`

### ISSUE THREAD

**Issue by byjiang1996** [flashinfer-ai/flashinfer#1202]
Hello FlashInfer team!

I tried to replace `torch.div&torch.softmax` to this flashinfer fuse kernel in sglang: https://github.com/flashinfer-ai/flashinfer/pull/1153. [sglang code changes here](https://github.com/byjiang1996/sglang/commit/da5a65e76749e86cc8e74200fd5a08a55c2a7525)

However, in both cases, inference performance degrades after moving to the fuse kernel. Do you have any idea why?

## Before: via pytorch native
model sampler's div&softmax via pytorch native: 45us
<img width="998" alt="image" src="https://github.com/user-attachments/assets/324706c1-2601-4b4d-a834-dc367c09eefa" />

spec decoding verify's div&softmax via pytorch native: 28us
<img width="1220" alt="image" src="https://github.com/user-attachments/assets/29ed555f-dc5a-4c06-b7f7-fe073cf5e194" />

## After: via fuse kernel
model sampler's div&softmax via fuse kernel: 92us
<img width="680" alt="image" src="https://github.com/user-attachments/assets/7ecd7a90-bf18-4f92-a1a1-9d444b0d454d" />

spec decoding verify's div&softmax via fuse kernel: 69us
<img width="945" alt="image" src="https://github.com/user-attachments/assets/6ec96e2e-c4bc-46ed-8763-23d85783e45c" />

## Reproduce
* [sglang code changes here](https://github.com/byjiang1996/sglang/commit/da5a65e76749e86cc8e74200fd5a08a55c2a7525)
* sglang server: `python3 -m sglang.launch_server --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 --attention-backend fa3 --trust-remote-code --dtype float16 --speculative-algo EAGLE3 --model-path /shared/public/models/meta-llama/Meta-Llama-3.1-8B-Instruct --speculative-draft-model-path /shared/public/elr-models/jamesliu1/sglang-EAGLE3-Llama-3.1-Instruct-8B/e5ed08d66f528a95ce89f5d4fd136a28f6def714`
* sglang client: `python3 -m sglang.bench_serving --backend sglang --dataset-name random --num-prompts 16 --sharegpt-output-len 100 --extra-request-body '{"top_k": 4, "temperature": 1.0}' --profile`

**Comment by yzh119** [comment]
Hi @byjiang1996 what's the problem shape here `(num_tokens, vocab_size)`

**Comment by byjiang1996** [comment]
Hi @yzh119 , thanks for your quick response!

During my benchmark
* model sampler's div&softmax: input logits shape is (1, 128256)
* spec decoding verify's div&softmax: input logits shape is (4, 128256)

**Comment by yzh119** [comment]
Hi @byjiang1996 can you provide a minimal reproducible script?

**Comment by byjiang1996** [comment]
Hi @yzh119 , thanks for helping here! You can reproduce by doing this:

## Setup
```
git clone git@github.com:byjiang1996/sglang.git
cd sglang
git checkout byjiang1996/flashinfer_softmax
python3 -m venv venv
source venv/bin/activate
pip3 install -e "python[all]" --index-url=https://pypi.org/simple
export PYTHONPATH=$(pwd)/python:$PYTHONPATH
```

## Benchmark after kernel fusing

Server:
```
python3 -m sglang.launch_server --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 --attention-backend fa3 --trust-remote-code --dtype float16 --speculative-algo EAGLE3 --model-path /shared/public/models/meta-llama/Meta-Llama-3.1-8B-Instruct --speculative-draft-model-path /shared/public/elr-models/jamesliu1/sglang-EAGLE3-Llama-3.1-Instruct-8B/e5ed08d66f528a95ce89f5d4fd136a28f6def714
```

After server is ready, launch client

Client: 
```
python3 -m sglang.bench_serving --backend sglang --dataset-name random --num-prompts 16 --sharegpt-output-len 100 --extra-request-body '{"top_k": 4, "temperature": 1.0}' --profile
```

After client terminates, you will find a .json.gz profile file under /tmp folder

This is the profile file i generated before: [new_softmax.json.gz](https://github.com/user-attachments/files/21023614/new_softmax.json.gz). You can visualize it in https://ui.perfetto.dev/
 
## Benchmark before kernel fusing
```
git reset --hard head~ ## to revert the kernel fuse commit
```

Server:
```
python3 -m sglang.launch_server --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 --attention-backend fa3 --trust-remote-code --dtype float16 --speculative-algo EAGLE3 --model-path /shared/public/models/meta-llama/Meta-Llama-3.1-8B-Instruct --speculative-draft-model-path /shared/public/elr-models/jamesliu1/sglang-EAGLE3-Llama-3.1-Instruct-8B/e5ed08d66f528a95ce89f5d4fd136a28f6def714
```

After server is ready, launch client

Client: 
```
python3 -m sglang.bench_serving --backend sglang --dataset-name random --num-prompts 16 --sharegpt-output-len 100 --extra-request-body '{"top_k": 4, "temperature": 1.0}' --profile
```

After client terminates, you will find a .json.gz profile file under /tmp folder

This is the profile file i generated before: [old_softmax.json.gz](https://github.com/user-attachments/files/21023643/old_softmax.json.gz). You can visualize it in https://ui.perfetto.dev/
